pkg/store/proxy.go

```diff
@@ -269,10 +331,40 @@ func startStreamSeriesSet(
 	return s
 }
 
-// Next blocks until new message is received or stream is closed.
+// Next blocks until new message is received or stream is closed or operation is timed out.
 func (s *streamSeriesSet) Next() (ok bool) {
-	s.currSeries, ok = <-s.recvCh
-	return ok
+	ctx := s.ctx
+	timeoutMsg := ""
+
+	if s.readTimeout != 0 {
+		timeoutMsg = s.readTimeout.String()
+
+		timeoutCtx, done := context.WithTimeout(s.ctx, s.readTimeout)
+		defer done()
+		ctx = timeoutCtx
+	}
+
+	select {
+	case s.currSeries, ok = <-s.recvCh:
+		return ok
+	case <-ctx.Done():
+		//shutdown a goroutine in startStreamSeriesSet
+		s.closeSeries()
+
+		err := errors.Wrapf(ctx.Err(), "failed to receive any data in %s from %s", timeoutMsg, s.name)
```

Comment:
If `s.readTimeout` is equal to 0 then we will print here something like:
`failed to receive any data in from 1.2.3.4:1234`
Right? Could we improve this error message somehow?